src/components/molecules/OakModal/OakModal.tsx

```diff
@@ -0,0 +1,196 @@
+import { FocusOn } from "react-focus-on";
+import { Transition, TransitionStatus } from "react-transition-group";
+import styled from "styled-components";
+
+import { InternalShadowRoundButton } from "../InternalShadowRoundButton";
+
+import { OakBox, OakFlex, OakImage } from "@/components/atoms";
+import { parseOpacity } from "@/styles/helpers/parseOpacity";
+
+export type OakModalProps = {
+  /**
+   * The content of the modal.
+   * Use with `<OakModalBody>` for best results.
+   */
+  children: ReactNode;
+  /**
+   * Slot for the footer of the modal.
+   * Use with `<OakModalFooter>` for best results.
+   */
+  footerSlot?: ReactNode;
+  /**
+   * Indicates whether the modal is open or closed
+   */
+  isOpen: boolean;
+  /**
+   * Called when the modal is closed
+   */
+  onClose: () => void;
+  /**
+   * The DOM container to render the modal portal into.
+   *
+   * @default document.body
+   */
+  domContainer?: Element;
+} & Pick<
+  HTMLAttributes<Element>,
+  "aria-label" | "aria-description" | "aria-labelledby" | "aria-describedby"
+>;
+
+const FadeOutBox = styled(OakBox)<{ $state: TransitionStatus }>`
+  opacity: ${({ $state }) => {
+    switch ($state) {
+      case "entered":
+      case "entering":
+        return parseOpacity("semi-transparent");
+      default:
+        return parseOpacity("transparent");
+    }
+  }};
+`;
+
+const SlideInFlex = styled(OakFlex)<{ $state: TransitionStatus }>`
+  transform: ${({ $state }) => {
+    switch ($state) {
+      case "entered":
+      case "entering":
+        return "translateX(0)";
+      default:
+        return "translateX(-100%)";
+    }
+  }};
+`;
+
+const logoSrc = `https://${process.env.NEXT_PUBLIC_OAK_ASSETS_HOST}/${process.env.NEXT_PUBLIC_OAK_ASSETS_PATH}/logo-mark.svg`;
+
+/**
+ * Modal dialog with trapped focus and a close button.
+ */
+export const OakModal = ({
+  children,
+  footerSlot,
+  domContainer = document.body,
+  isOpen,
+  onClose,
+  ...rest
+}: OakModalProps) => {
+  const [canaryElement, setCanaryElement] = useState<HTMLDivElement | null>(
+    null,
+  );
+  const [isScrolled, setIsScrolled] = useState(false);
+  const transitionRef = useRef<HTMLDivElement>(null);
+
+  useLayoutEffect(() => {
+    if (!canaryElement) {
+      return;
+    }
+    const observer = new IntersectionObserver(
+      (mutations) => {
+        setIsScrolled(!mutations.some((mutation) => mutation.isIntersecting));
+      },
+      {
+        root: canaryElement.parentElement,
+      },
+    );
+    observer.observe(canaryElement);
+
+    return () => {
+      observer.disconnect();
+    };
+  }, [canaryElement]);
+
+  return createPortal(
+    <Transition
+      in={isOpen}
+      nodeRef={transitionRef}
+      addEndListener={(done) => {
+        transitionRef.current?.addEventListener("transitionend", done);
+      }}
+      timeout={500}
+      mountOnEnter
+      unmountOnExit
+    >
+      {(state) => (
+        <FocusOn onEscapeKey={onClose} returnFocus autoFocus>
```

Comment:
`returnFocus` will restore focus to the element that was active before the modal was opened (most likely the button that opened the modal in the first place) this ensures that the user does not lose their place in the document

`autoFocus` will give focus to a focusable element inside the modal — typically the first element in the tab index.